apps/hash-ai-worker-ts/src/activities/flow-activities/answer-question-action.ts

```diff
@@ -0,0 +1,448 @@
+
+  if (contextToUpload) {
+    const requestId = Context.current().info.workflowExecution.workflowId;
+    await sandbox.uploadFile(Buffer.from(contextToUpload), requestId);
+  }
+
+  const response = await sandbox.runPython(code);
+
+  logger.debug(`Python code execution response: ${stringify(response)}`);
+
+  const { stdout, stderr, artifacts } = response;
+
+  const downloadedArtifacts = await Promise.all(
+    artifacts.map(async (artifact) => {
+      const rawArtifact = await artifact.download();
+      return rawArtifact;
+    }),
+  );
+
+  await sandbox.close();
+
+  return { stdout, stderr, artifacts: downloadedArtifacts };
+};
+
+const systemPrompt = dedent(`
+  You are an expert data analyst. You will be provided with data to analyze, which may be in the form of text,
+  a graph of structured entities with links between them, a spreadsheet, or some combination of these.
+  
+  Your boss will ask you to answer a question based on the data provided to you. They might have asked for a specific output format,
+  e.g. 'Markdown', 'JSON', or 'CSV'. If it isn't specified, you use your best judgment.
+  
+  Your boss plans to use your answer to make a decision, so you make sure that it is accurate, concise, and clear.
+  If you can''t provide an answer based on the available data, you explain why, and request more data if it would help.
+  
+  You write Python to analyze any context provided, if you need it. In your code, you access the context from a file using the provided path.
+  Your Python code contains detailed comments about why each function is used, and how the data is accessed, including references to the shape of the context data.
+  You code complete the entire task in one file.
+  
+  You provide a confidence score with your answer, which is a number between 0 and 1. 
+  1 represents absolute certainty, and 0 a complete guess – you never provide answers with confidence 0, but instead explain why you can't answer.
+`);
+
+type ModelResponseArgs = {
+  answer?: string;
+  explanation: string;
+  confidence?: number;
+  code?: string;
+};
+
+const callModel = async (
+  messages: OpenAI.ChatCompletionCreateParams["messages"],
+  context: string | null,
+  codeUsed: string | null,
+): Promise<
+  Status<{
+    outputs: StepOutput[];
+  }>
+> => {
+  const openApiPayload: OpenAI.ChatCompletionCreateParams = {
+    model: "gpt-4-turbo-preview",
+    messages,
+    stream: false,
+    temperature: 0,
+    tools: answerTools,
+  };
+
+  const openAiResponse = await getOpenAiResponse(openApiPayload);
+
+  logger.debug(`Open AI Response received: ${stringify(openAiResponse)}`);
+
+  const { response } = openAiResponse.contents[0]!;
+
+  const { message: modelResponseMessage } = response;
+
+  const toolCalls = modelResponseMessage.tool_calls;
+
+  if (!toolCalls?.length) {
+    return {
+      code: StatusCode.Internal,
+      message: "No tool call found in response",
+      contents: [],
+    };
+  }
```

Comment:
We could also use a retry message here instead

Reply:
Added in https://github.com/hashintel/hash/pull/4290/commits/61e269b58f86ba46ac83bd50f876535c406c840c, as well as retrying if no valid tool call is found